### DISCUSSION TITLE
usb keyboard (berry)

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#22495] [Q&A]
is there any way to get input from usb Keyboard ?
i have on of those boards
![20739_3](https://github.com/user-attachments/assets/c7facb11-b5ed-4848-b744-4cc8854ce1d9)

**Comment by sfromis** [comment]
No "full" support for USB, only for USB CDC for serial over USB.

**Reply by barbudor** [reply]
On ESP32-S3, the USB support OTG which means it can act either as a device (when you connect it to a host like your PC) or as a host (to connect a device).
So in theory it should be possible with an ESP32-S3 and there might be Arduino library to support USB Host mode and a keyboard.
Nevertheless, Tasmota does not include such library/driver. That would require some work for a very little edge use-case.

**Comment by jpmorrison** [comment]
CH9329 or 9350 takes care of all the USB HID part, you just need a uart.
I’m getting some to do some wireless kvm and for a USB UPS.
